### DISCUSSION TITLE
Suggestion on feature/scenario/step organisation

### DISCUSSION THREAD

**Discussion by tngok-wustl** [Tyler-Keith-Thompson/CucumberSwift#93] [Q&A]
Hi! I have managed to set up CucumberSwift BDD based on your tutorial steps. Now, I want to incorporate CucumberSwift to my project's unit test, which tests navigation between screens (view controllers). So, how would you suggest me organising the steps? Should I create one single features with multiple scenarios, or multiple features with one scenario for each, or something else? Thanks!

**Comment by Tyler-Keith-Thompson** [comment]
Well, there's a few choices. If you're looking at unit tests, [the DSL](https://github.com/Tyler-Keith-Thompson/CucumberSwift/wiki/Gherkin-DSL-for-Swift) might be worth it. This is assuming you didn't want a separate feature file. This allows for very flexible code organization and could drive potentially different feature/scenario patterns.

If that doesn't work for you, my advice is to stick with whatever seems readable. I'd imagine you would want a feature for cohesive sets of flows in your app. Perhaps a feature for all the navigation related to <common user flow X> and multiple scenarios describing what happens within that.

**Reply by tngok-wustl** [reply]
Thank you for your response. I am now writing CucumberSwift UI tests for a simple two-scenario navigation (i.e., given a button exists on a screen, clicking on the button and then go to the next screen).

Here is my .feature file:
```
Feature: Basic UI Tests

    Scenario: Open the upload menu
        Given upload button exists
        When clicking the upload button
        Then upload menu opened

    Scenario: Go to create a recipe title
        Given upload recipe button exists
        When clicking the upload recipe button
        Then create recipe title screen opened
```

And here is my CucumberSwift UI test script:
```
import XCTest
import CucumberSwift

extension Cucumber: StepImplementation {
    public var bundle: Bundle {
        class CucumberUITest: CucumberTest { }
        return Bundle(for: CucumberUITest.self)
    }
    
    public func setupSteps() {
        let app = XCUIApplication()
        
        BeforeFeature { _ in
            print("CucumberSwift: Launching app...")
            app.launch()
        }
        
        Feature("Basic UI Tests") {
            Scenario("Open the upload menu") {
                Given(/^upload button exists$/) { _, _ in
                    XCTAssert(app.buttons["uploadButton"].exists)
                }
                
                When(/^clicking the upload button$/) { _, _ in
                    app.buttons["uploadButton"].tap()
                }
                
                Then(/^upload menu opened$/) { _, _ in
                    XCTAssert(app.staticTexts["Select Upload"].exists)
                }
            }
            
            Scenario("Go to create a recipe title") {
                Given(/^upload recipe button exists$/) { _, _ in
                    XCTAssert(app.buttons["uploadRecipe"].exists)
                }
                
                When(/^clicking the upload recipe button$/) { _, _ in
                    app.buttons["uploadRecipe"].tap()
                }
                
                Then(/^create recipe title screen opened$/) { _, _ in
                    XCTAssert(app.staticTexts["Creating Recipe"].exists)
                    XCTAssert(app.staticTexts["Recipe Name"].exists)
                }
            }
        }
        
        AfterFeature { _ in
            print("CucumberSwift: Terminating app...")
            app.terminate()
        }
    }
}
```

However, the UI tests are behaving weirdly (i.e., not as what I expect):
1. The second scenario runs before the first one (I would expect the first scenario to run first, because the first scenario is higher up in the .feature file)
2. The «Then» step runs before the «When» step (I would expect the step running order to be Given > When > Then)
3. App relaunches between scenarios (I would expect the app to launch only before the feature, and there is only one feature)
4. Each step is run twice

So, could you please tell me what is wrong with my code and explain the weird behaviours above one by one?

**Reply by beaubrownsdet** [reply]
I am also experiencing the "Then" scenario running before the "When" scenario. Did you ever figure this out, @tngok-wustl?

**Reply by beaubrownsdet** [reply]
I was able to get the tests to run in order by disabling parallel test execution.
